File: EmployeeManagementSystem - backend/Services/ReportService.cs
using EmployeeManagementSystem.Models;
using EmployeeManagementSystem.Repositories;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Threading.Tasks;

namespace EmployeeManagementSystem.Services
{
    public class ReportService: IReportService
    {
        private readonly IReportRepository _reportRepository;
        private readonly IEmployeeRepository _employeeRepository;

        public ReportService(IReportRepository reportRepository, IEmployeeRepository employeeRepository)
        {
            _reportRepository = reportRepository;
            _employeeRepository = employeeRepository;
        }

        public async Task<ReportDto> GetEmployeeDirectoryAsync()
        {
            var employees = await _reportRepository.GetAllEmployeesWithDetailsAsync();

            var employeeData = employees.Select(e => new
            {
                Id = e.Id,
                Name = $"{e.FirstName} {e.LastName}",
                Email = e.Email,
                Phone = e.Phone,
                Department = e.Department,
                Position = e.Position,
                Manager = e.Manager != null ? $"{e.Manager.FirstName} {e.Manager.LastName}" : "None",
                JoinDate = e.JoinDate.ToString("yyyy-MM-dd"),
                IsActive = e.IsActive
            }).ToList();

            return new ReportDto
            {
                ReportType = "Employee Directory",
                GeneratedDate = DateTime.Now,
                Data = employeeData
            };
        }

        public async Task<ReportDto> GetLeaveAnalysisAsync(DateTime? startDate = null, DateTime? endDate = null)
        {
            var leaveRequests = await _reportRepository.GetAllLeaveRequestsAsync(startDate, endDate);

            var leaveData = leaveRequests.Select(lr => new
            {
                EmployeeName = $"{lr.Employee.FirstName} {lr.Employee.LastName}",
                Department = lr.Employee.Department,
                LeaveType = lr.LeaveType,
                StartDate = lr.StartDate.ToString("yyyy-MM-dd"),
                EndDate = lr.EndDate.ToString("yyyy-MM-dd"),
                Duration = (lr.EndDate - lr.StartDate).Days + 1,
                Reason = lr.Reason,
                Status = lr.Status,
                ApprovedBy = lr.ApprovedBy != null ? $"{lr.ApprovedBy.FirstName} {lr.ApprovedBy.LastName}" : "N/A"
            }).ToList();

            return new ReportDto
            {
                ReportType = "Leave Analysis",
                GeneratedDate = DateTime.Now,
                Data = leaveData
            };
        }

        public async Task<ReportDto> GetDepartmentDistributionAsync()
        {
            var distribution = await _reportRepository.GetEmployeesByDepartmentAsync();

            return new ReportDto
            {
                ReportType = "Department Distribution",
                GeneratedDate = DateTime.Now,
                Data = distribution
            };
        }

        public async Task<ReportDto> GetLeaveDistributionByTypeAsync(DateTime? startDate = null, DateTime? endDate = null)
        {
            var distribution = await _reportRepository.GetLeaveStatisticsByTypeAsync(startDate, endDate);

            return new ReportDto
            {
                ReportType = "Leave Distribution by Type",
                GeneratedDate = DateTime.Now,
                Data = distribution
            };
        }

        public async Task<ReportDto> GetAverageLeavesByDepartmentAsync(int year)
        {
            var averages = await _reportRepository.GetAverageLeavesByDepartmentAsync(year);

            return new ReportDto
            {
                ReportType = $"Average Leaves by Department ({year})",
                GeneratedDate = DateTime.Now,
                Data = averages
            };
        }

        public async Task<ReportDto> GetManagerHierarchyAsync()
        {
            // This is a bonus report leveraging your self-referencing manager relationship
            var employees = await _reportRepository.GetAllEmployeesWithDetailsAsync();

            // Find top-level managers (those without managers)
            var topManagers = employees.Where(e => e.ManagerId == null && e.IsActive).ToList();

            // Build hierarchy
            var hierarchy = topManagers.Select(m => BuildManagerHierarchy(m, employees)).ToList();

            return new ReportDto
            {
                ReportType = "Manager Hierarchy",
                GeneratedDate = DateTime.Now,
                Data = hierarchy
            };
        }

        private object BuildManagerHierarchy(Employee manager, IEnumerable<Employee> allEmployees)
        {
            var subordinates = allEmployees.Where(e => e.ManagerId == manager.Id && e.IsActive).ToList();

            return new
            {
                Name = $"{manager.FirstName} {manager.LastName}",
                Position = manager.Position,
                Department = manager.Department,
                Subordinates = subordinates.Select(s => BuildManagerHierarchy(s, allEmployees)).ToList()
            };
        }
    }
}

File: EmployeeManagementSystem - frontend/src/services/reportService.js



import api from './api';

export const reportService = {
  getEmployeeDirectory: async () => {
    try {
      console.log("Fetching employee directory...");
      const response = await api.get('/reports/employee-directory');
      console.log("Employee directory response:", response.data);
      return response.data;
    } catch (error) {
      console.error("Error fetching employee directory:", error.response?.data || error.message);
      throw error;
    }
  },

  getLeaveAnalysis: async (startDate, endDate) => {
    try {
      let url = '/reports/leave-analysis';
      
      // Add query parameters if provided
      if (startDate || endDate) {
        url += '?';
        if (startDate) {
          url += `startDate=${startDate.toISOString()}`;
        }
        if (endDate) {
          url += startDate ? `&endDate=${endDate.toISOString()}` : `endDate=${endDate.toISOString()}`;
        }
      }
      
      console.log(`Fetching leave analysis with URL: ${url}`);
      const response = await api.get(url);
      console.log("Leave analysis response:", response.data);
      return response.data;
    } catch (error) {
      console.error("Error fetching leave analysis:", error.response?.data || error.message);
      throw error;
    }
  },

  getDepartmentDistribution: async () => {
    try {
      console.log("Fetching department distribution...");
      const response = await api.get('/reports/department-distribution');
      console.log("Department distribution response:", response.data);
      return response.data;
    } catch (error) {
      console.error("Error fetching department distribution:", error.response?.data || error.message);
      throw error;
    }
  },

  getLeaveDistributionByType: async (startDate, endDate) => {
    try {
      let url = '/reports/leave-distribution-by-type';
      
      // Add query parameters if provided
      if (startDate || endDate) {
        url += '?';
        if (startDate) {
          url += `startDate=${startDate.toISOString()}`;
        }
        if (endDate) {
          url += startDate ? `&endDate=${endDate.toISOString()}` : `endDate=${endDate.toISOString()}`;
        }
      }
      
      console.log(`Fetching leave distribution by type with URL: ${url}`);
      const response = await api.get(url);
      console.log("Leave distribution response:", response.data);
      return response.data;
    } catch (error) {
      console.error("Error fetching leave distribution:", error.response?.data || error.message);
      throw error;
    }
  },

  getAverageLeavesByDepartment: async (year) => {
    try {
      console.log(`Fetching average leaves by department for year ${year}`);
      const response = await api.get(`/reports/average-leaves-by-department/${year}`);
      console.log("Average leaves response:", response.data);
      return response.data;
    } catch (error) {
      console.error("Error fetching average leaves:", error.response?.data || error.message);
      throw error;
    }
  },

  getManagerHierarchy: async () => {
    try {
      console.log("Fetching manager hierarchy...");
      const response = await api.get('/reports/manager-hierarchy');
      console.log("Manager hierarchy response:", response.data);
      return response.data;
    } catch (error) {
      console.error("Error fetching manager hierarchy:", error.response?.data || error.message);
      throw error;
    }
  }
};

export default reportService;